cabal.project

```diff
@@ -48,3 +48,16 @@ source-repository-package
   subdir:
     fs-api
     fs-sim
+
+-- This relaxing of dependency bounds on @base@ is currently required in order to
+-- build the Haddock documentation with the @scripts/haddock.sh@ script.
+allow-newer:
+  fs-api:base,
+  fs-sim:base,
+  io-classes:base,
+  io-sim:base,
+  quickcheck-lockstep:base,
+  si-timers:base,
+  strict-mvar:base,
+  strict-stm:base,
+  unix-bytestring:base,
```

Comment:
Let's make this conditional on the GHC version

```suggestion
-- TODO: this relaxing of dependency bounds on @base@ is currently required in order to
-- build the Haddock documentation with the @scripts/haddock.sh@ script.
if impl(ghc >=9.10)
  allow-newer:
    fs-api:base,
    fs-sim:base,
    io-classes:base,
    io-sim:base,
    quickcheck-lockstep:base,
    si-timers:base,
    strict-mvar:base,
    strict-stm:base,
    unix-bytestring:base,
```

Reply:
> Let's make this conditional on the GHC version

I forgot the `cabal.project` format permitted the conditional keywords. This is a good catch!

Still, I'll see if I can remove the need for most (hopefully all) of the dependencies in the `allow-newer` stanza by tweaking their upper bounds in the package description file.

Reply:
> I'd also suggest this change to .github/workflows/haskell.yml to ensure our CI also builds/tests with ghc-9.10.1

I had refrained from updating the CI pipeline because I wasn't sure if the branch would be culled and didn't want to allocate excessive (probably overly cautious) resources in the CI pipeline. Since the changes are well received this makes sense. Thanks!

Reply:
FYI, you probably don't have to update many bounds in the package description file, just the `index-state` stanzas (see https://github.com/IntersectMBO/lsm-tree/pull/319#pullrequestreview-2213903964)

Reply:
I think it turned out to be a little bit more involved (see the new changes) than the index state since neither `fs-api` nor `fs-sim` have a version which supports GHC 9.10. I'm not sure why, but the updated `index-state` is not having the [IntersectMBO/cardano-haskell-packages](https://github.com/IntersectMBO/cardano-haskell-packages) repository forward the GHC 9.10 compatible `fs-api` source code from the [input-output-hk/fs-sim](https://github.com/input-output-hk/fs-sim) repository. I would like to see how much of the `fs-api:*` and `fs-sim:*` items can be removed from `allow-newer` if I can get the newest source code, but I'm not savvy enough to navigate the plumbing of the TOML files and cabal revisions.

Reply:
Never-mind, I didn't notice that the `fs-api` and `fs-sim` versions were explicitly pinned to a commit hash. I can update the commit hash to a newer version which supports GHC 9.10.